test: Check connecting with known_hosts entry

martinpitt opened the pull request:
Test both with a correct matching, and a changed key.  Both scenarios fail right now, as for the "good" case, ferny asks for confirming the  host key even if it is already known and correct; and for the "bad" case, ferny currently entirely ignores the existing global and user known_hosts files, thus is unable to actually detect a changed host key.
    
See https://github.com/cockpit-project/cockpit/issues/18713

-----

I have a vague idea how to fix this, but I'd like to first discuss the desired behaviour with @allisonkarlitskaya -- i.e. pin down the unit tests. I now did enough investigation in the cockpit issue that I believe I understand what cockpit is expecting.

martinpitt commented:
Meh, [what's up with the C8S repos](https://github.com/allisonkarlitskaya/ferny/actions/runs/5083299280/jobs/9134673887?pr=15)? But anyway, I just realized that we don't test the `handle_host_key=False` case at all, let's do that.

martinpitt commented:
*grmbl* old Python 3.6 *mumble* *mumble*

allisonkarlitskaya commented on src/ferny/errors.py:
```diff
@@ -90,7 +94,8 @@ def make_gaierror_map() -> Iterable[Tuple[str, int]]:
 def get_exception_for_ssh_stderr(stderr: str) -> Exception:
-    for ssh_cls in [AuthenticationError, HostKeyError]:
+    # ChangedHostKeyError must come first, as the HostKeyError message is also present in the "key has changed" error
+    for ssh_cls in [AuthenticationError, ChangedHostKeyError, HostKeyError]:
```
Could we change the other message so that the conditions are not overlapping and the order is irrelevant?  Otherwise I'd prefer something slightly more formal for ordering ­— maybe lifting this out to a top-level constant for example.  I completely missed the comment here on first read.

martinpitt replied:
For "changed" it looks like this:
```
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@    WARNING: REMOTE HOST IDENTIFICATION HAS CHANGED!     @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
IT IS POSSIBLE THAT SOMEONE IS DOING SOMETHING NASTY!
Someone could be eavesdropping on you right now (man-in-the-middle attack)!
It is also possible that a host key has just been changed.
The fingerprint for the RSA key sent by the remote host is
SHA256:ZeBykpPEufJoNUuDWQTEnA+pIdkQUralwtLLItxK+GM.
Please contact your system administrator.
Add correct host key in /tmp/pytest-of-martin/pytest-103/test_no_host_key_changed0/keys/known_hosts to get rid of this message.
Offending RSA key in /tmp/pytest-of-martin/pytest-103/test_no_host_key_changed0/keys/known_hosts:1
Host key for [127.0.0.1]:53187 has changed and you have requested strict checking.
Host key verification failed.
```

For "unknown" it looks like this:
```
No RSA host key is known for [127.0.0.1]:51609 and you have requested strict checking.
Host key verification failed.
```
So the "No .* key is known" would provide a non-overlapping match. We could try to match on the entire message instead of just a single line, and include the whole thing in the error message.

martinpitt replied:
Done, and I split off `UnknownHostKey` to make this more symmetric and safer to use.

martinpitt replied:
Ah dang, [RHEL 8 can't do that](https://github.com/allisonkarlitskaya/ferny/actions/runs/5088646507/jobs/9145312889?pr=15). I'll need to rework that using conditional xfails, and a RHEL 8 specific test perhaps.

martinpitt replied:
I reworked this to keep the generic error as a fallback. Unfortunately this means to bring back the order dependence, I updated the comment.

allisonkarlitskaya reviewed (commented):
Looks good, thanks for working on this!  I had a feeling that the one regex grabbing both situations was the cause of the issue.

I'm also surprised that the hostkey callback gets called.  That should be literally impossible.  Are you sure that's not caused by the fallback code in the responder that takes things that look like host key validation requests in do_askpass() and shunts them over to the hostkey function?  If so, cleaning it up there would be better than an xfail.

martinpitt commented:
> I'm also surprised that the hostkey callback gets called. That should be literally impossible.

Yeah, I didn't expect that either.

> Are you sure that's not caused by the fallback code in the responder that takes things that look like host key validation requests in do_askpass() and shunts them over to the hostkey function

It's related to that. It calls `do_askpass` with this prompt:

```
The authenticity of host '[127.0.0.1]:46039 ([127.0.0.1]:46039)' can't be established.
RSA key fingerprint is SHA256:ZeBykpPEufJoNUuDWQTEnA+pIdkQUralwtLLItxK+GM.
This key is not known by any other names
Are you sure you want to continue connecting (yes/no/[fingerprint])?
```

So in that sense, handing it over to `do_hostkey` is correct, but I understand this was only meant as a fallback for RHEL 8, and should not actually happen on Fedora? I.e. this indicates an actual bug somewhere. _askpass_task() gets called with *two* arguments, which it interprets as password prompt (hostkey prompt is 6):

```
DEBUG    ferny.interaction_agent:interaction_agent.py:132 _askpass_task(['/run/user/1000/ferny/tmpf3ve8ra_/ferny-askpass', "The authentici
ty of host '[127.0.0.1]:59663 ([127.0.0.1]:59663)' can't be established.\nRSA key fingerprint is SHA256:ZeBykpPEufJoNUuDWQTEnA+pIdkQUralwt
LLItxK+GM.\nThis key is not known by any other names\nAre you sure you want to continue connecting (yes/no/[fingerprint])? "], 52, <_io.Te
xtIOWrapper name=13 mode='w' encoding='UTF-8'>, <_io.TextIOWrapper name=14 mode='w' encoding='UTF-8'>, )
```

martinpitt commented:
I added another commit which checks the agent callback arguments.

allisonkarlitskaya commented:
> That feels wrong to me. _askpass_task() could perhaps intercept that and raise a HostKeyError straight away, without calling the agent?

I'm not sure how I feel about this.

I start to prefer the "traditional way" of dealing with host key prompts (ie: the usual TOFU) thing.  That's not how cockpit currently works, so that's not how ferny developed, but I think in the future it would be meaningful to have ssh waiting for the user to interact with the "add host key?" dialog in the browser before writing either "yes" or "no" back to the askpass response.

That's markedly *not* what the current hostkeys callback is about.  It has no ability to either:
 - check if the key is already in some file
 - add a key to the file on accepting it

At the same time, though, if our intended way of accepting host keys in the future is by writing `yes` at the TOFU prompt, we're going to need to develop some kind of a set of regular expression for identifying those prompts.  I think it makes sense to do that inside of ferny, but probably doesn't make sense to do it with the existing API... If that makes sense... :)

martinpitt commented:
@allisonkarlitskaya I reworked `test_no_host_key_unknown` according to the updated expectation when do_askpass() happens. It's not xfail any more, and I added some commentary what happens.

martinpitt commented:
Reworked again to make test_known_host_changed() pass. That was a wrong assumption on my part. I initially thought that ferny should just reject changed host keys without prompting, but (1) it does, and (2) cockpit also actually needs that.

allisonkarlitskaya commented on test/test_basic.py:
```diff
@@ -14,6 +14,13 @@
 os.environ.pop('SSH_ASKPASS', None)
 
+# some host key which isn't the one from mock-ssh
+NONMATCHING_HOSTKEY = (
```
I feel like maybe this should be in a file.

At the same time, I can't find the officially suggested way to do that with pytest, and I don't really feel like forcing you to do another round...

allisonkarlitskaya reviewed (approved):
I'm happy, thanks.

Let's get this landed.  I want to start exploring what the new API might look like from the ferny side.

martinpitt commented on test/test_basic.py:
```diff
@@ -158,6 +175,31 @@ async def test_wrong_passphrase(self, key_dir: pathlib.Path, runtime_dir: pathli
+    @pytest.mark.xfail
+    async def test_known_host_good(self, key_dir: pathlib.Path, runtime_dir: pathlib.Path) -> None:
+        await self.run_test(key_dir, runtime_dir, ZeroDivisionError(), 'passphrase',
+                            known_host_key='scan')
```
This is really annoying. When doing this interactively, this works just fine:
```
❱❱❱ ssh -o UserKnownHostsFile=/tmp/pytest-of-martin/pytest-200/test_known_host_good0/keys/known_hosts -i /tmp/pytest-of-martin/pytest-200/test_known_host_good0/keys/id_rsa -p 39597 -l admin -o StrictHostKeyChecking=yes 127.0.0.1 hostname
cockpit-toolbox
```

but when I create an echo-only KnownHostsCommand:
```
❱❱❱ ssh -o UserKnownHostsFile=/tmp/pytest-of-martin/pytest-200/test_known_host_good0/keys/known_hosts -i /tmp/pytest-of-martin/pytest-200/test_known_host_good0/keys/id_rsa -p 39597 -l admin -o 'KnownHostsCommand=/tmp/xx %I %H %t %K %f' -o CheckHostIP=yes -o StrictHostKeyChecking=yes 127.0.0.1 hostname
YYY KnownHostsCommand ORDER 127.0.0.1 NONE NONE NONE
YYY KnownHostsCommand HOSTNAME [127.0.0.1]:39597 ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDWdxkgPs4niiaW41P8NiKjI3slCoeaRQvvchTHCyvQMGOanv+iudgurkcVvJOWOHsbLdxSfW5KbF1bGpVu3nwjbA7rajDx8Xs4z6VLsd4WCrHJl0qZt5GFfYTriIiPfE1t/C9MIxA1Vfxz099DBQDgs96kt7EidP2cBTb1rWGjBAt71jlfuxH4g1+emuPcuhdY3PFH5Ac7IwG5So3jxUWB7esDiO7StoKcAU2iJzp8yFLOrekYn8IA9cAxyzgYzlnqs8S/6aFrm/xTYAb2YIGyLUoyQgAIQW4MlILxq5opS5+YUYZaBLZRYoI3vkqqF+ULeqdZzgcOSLe4cbE3bZql SHA256:ZeBykpPEufJoNUuDWQTEnA+pIdkQUralwtLLItxK+GM
cockpit-toolbox
```

i.e. it still calls the KnownHostsCommand with a `HOSTNAME` **even if** it already knows it:

```
debug1: Found key in /tmp/pytest-of-martin/pytest-200/test_known_host_good0/keys/known_hosts:1
[...]
debug1: Will attempt key: /home/martin/.ssh/id_rsa RSA SHA256:Eq3Xz7MwpEieWi+5do3xJ9l0b/VW4ryM1B84W399zfg agent
debug1: Will attempt key: martin@cockpit-toolbox RSA SHA256:tIilj2EYf3gVPkTWIYURax+sLjyCSoodlsKZ4nRQVfk agent
debug1: Will attempt key: /tmp/pytest-of-martin/pytest-200/test_known_host_good0/keys/id_rsa RSA SHA256:VCdgVD78sNf5IiV49xlTPYqqrITt8o+uGUFEtJmKfOw explicit
```
So ssh's behaviour there is a bit misleading. I suppose one could interpret the behaviour that way:

> Specifies a command to use to obtain a list of host keys, in addition to those listed in UserKnownHostsFile and GlobalKnownHostsFile.  This command is executed after the files have been read.  It may write host key lines to standard output in identical format to the usual files (described in the VERIFYING HOST KEYS section in ssh(1)).

martinpitt replied:
See PR #16
